### PR TITLE
[CPU] Internal dynamism support for extensions

### PR DESCRIPTION
**Pull request by maxnick**
### Details:
This PR introduce support for extension operations with so called internal dynamism, i.e. when the output shapes can only be defined after performing the computations, for example NMS operation.

### Tickets:
 - 99106
 - 81379

### CODE REVIEW

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference_status.hpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+namespace ov {
+namespace intel_cpu {
+
+enum class ShapeInferStatus {
+    update, // shapes were successfully calculated
```
Comment (praasz) [inline src/plugins/intel_cpu/src/utils/shape_inference/shape_inference_status.hpp:11]: `update` -> `success` as shape `infer` return value not update it?

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp [hunk 2]
```diff
@@ -20,8 +21,14 @@ void shape_inference(ov::Node* op,
...
 public:
-    virtual std::vector<StaticShape> infer(const std::vector<StaticShape>& input_shapes,
-                                           const std::map<size_t, HostTensorPtr>& constant_data) = 0;
+    struct ShapeInferResult {
+        std::vector<StaticShape> shapes;
+        ShapeInferStatus status;
+    };
+
+public:
+    virtual ShapeInferResult infer(const std::vector<StaticShape>& input_shapes,
```
Comment (praasz) [inline src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp:30]: As the infer result status has got only two states now (not sure if there is a plan to expand it), maybe there is a chance to use empty vector as skipped status?

In template implementation of shape infer it looks like it always returns not empty vector or throw exception if inference fails. The empty  vector could be interpreted as infer has been skipped. Then return type of this function doesn't need to be changed, and changes will be less impacting especially on classes where this status has no special meaning and they always return `updated`.

Other solution (to expand interface not modify it) can be add additional function to this interface e.g. `get_status()` which provide last infer status, for most classes it can be compile time value as it never changes.
Then `entryBase` can implement this getter with default (updated) value and only in required classes override it.

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference_cpu.hpp [hunk 3]
```diff
@@ -18,6 +20,11 @@ class IShapeInfer {
 public:
```
Comment (praasz) [inline src/plugins/intel_cpu/src/utils/shape_inference/shape_inference_cpu.hpp:20]: Maybe this interface can inherit from `IShapeInferCommon` (or some new Iface which got common members for both interfaces) or vice versa?

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp (hunk 2 above)
Comment (maxnick) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp:30]: Thank you for your suggestions.

>In template implementation of shape infer it looks like it always returns not empty vector or throw exception if inference fails. The empty vector could be interpreted as infer has been skipped. Then return type of this function doesn't need to be changed, and changes will be less impacting especially on classes where this status has no special meaning and they always return `updated`.

Of course, this is the first idea that came to my mind. But, what if the empty vector is a result of some error? In that case the program can not distinguish whether the empty vector is `skip` or we got unexpectedly empty data. Thus, we will have hardly detecting error. The sing is on the `IShapeInferCommon` hierarchy level we do not have control over the template `shape_infer` functions, and in the case of using an empty vector as a `skip` indication, we may just "misrepresent" an error as an valid result. But with the approach proposed in the PR we clearly indicate what we expect. Also, since this "complexity" is located on this level is think this is not a big problem.

>Other solution (to expand interface not modify it) can be add additional function to this interface e.g. get_status() which provide last infer status, for most classes it can be compile time value as it never changes.
Then `entryBase` can implement this getter with default (updated) value and only in required classes override it.

In fact, this approach does not reduce the complexity, but makes the interface less convenient, since the user should remember about the state that should be read with `get_status()`. Yes, in this proposal the user will also have to check the return status, but he receive it together with the resulting shape, so this is less likely that he forget to check it, so slightly more convenient.

But, to be frank, I do not completely like the solution either, but given the options we have, I think the proposed solution is better than others. Also in the CPU plugin I introduced the same `Result` type and it can be used to optimize the shape infer call. We can simply skip the computations if the input data have not been changed from the previous call.
If my reasoning is not good enough, further discussion is always welcome (:

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference_cpu.hpp (hunk 3 above)
Comment (maxnick) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference_cpu.hpp:20]: Unfortunately, we had to implement this interface, since the `IShapeInferCommon` use incompatible static shape representation and conversations take too much time. I mean `VectorDims` <-> `StaticShape`, map of host tensors <-> map of memory object pointers. So to implement optimized shape inference operations, optimized here means CPU plugin specifics and constraints, we introduced this interface, fully compatible with the plugin, and just wrapped `IShapeInferCommon` as more generic. But, I see that nobody else use `IShapeInferCommon` so m.b. the entire subsytem may adopt the CPU plugin specific `IShapeInfer `, but under the hood we will have to convert `VectorDims` <-> `StaticShape` and input maps anyway.

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference_status.hpp (hunk 1 above)
Comment (maxnick) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference_status.hpp]: Changed, thank you!

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp [hunk 4]
```diff
@@ -20,8 +21,14 @@ void shape_inference(ov::Node* op,
 
 class IShapeInferCommon {
 public:
-    virtual std::vector<StaticShape> infer(const std::vector<StaticShape>& input_shapes,
-                                           const std::map<size_t, HostTensorPtr>& constant_data) = 0;
+    struct ShapeInferResult {
```
[the remark this answers is not in the document]
Comment (maxnick) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp]: I renamed the struct, thank you for your suggestion!

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp (hunk 2 above)
Comment (praasz) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp]: Change interface is also fine, just was thinking about other solutions.
For me it looks like the `std::optional` from c++17 could be useful here.

**Review by praasz — APPROVED** [review]

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp (hunk 2 above)
Comment (maxnick) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp]: Completely agree with you. But we cannot use c++17 so far.

File: src/core/src/tensor_conversion_util.cpp [hunk 5]
```diff
@@ -18,7 +18,7 @@ Tensor wrap_tensor(const ngraph::HostTensorPtr& t) {
     } else if (p_shape.is_static()) {
         return {et, p_shape.to_shape(), t->get_data_ptr()};
     } else {
-        return {et, make_dynamic_shape()};
+        return {et, ov::Shape(p_shape.rank().get_length())};
```
Comment (praasz) [inline src/core/src/tensor_conversion_util.cpp:21]: Why is this changed?
It provides correct conversion between dynamic host tensor and ov::tensor in evaluates and  this WA fix a lot of issues until old tensor will be removed.


@ilyachur can you have a look?

**Review by praasz — CHANGES_REQUESTED** [review]

File: src/core/src/tensor_conversion_util.cpp (hunk 5 above)
Comment (maxnick) [reply src/core/src/tensor_conversion_util.cpp:21]: Obviously this changes do not work. Need to discuss working with `ov::Tensor`  in the case of dynamics shapes.
The main problem is that `set_shape` call to the `ov::Tensor` instance created from "dynamic" `ov::Shape`, which can not really represent a dynamic shape and that is why we use `ov::PartialShape`, can not be performed since in that case the tensor object is in the "uninitialized" state. But, what should the `evaluate` implementation do in that case? That is the question. Because following the interface of the `ov::Tensor` it seems that this is enough to simply call `set_shape`, but it turns out that it throws if the tensor is not initialized. In other words, `ov::Tensor` is not designed to have "undefined" state, but we need this to at least express operations with "internal" dynamism.

Comment (praasz) [reply src/core/src/tensor_conversion_util.cpp:21]: There is assumption that if `ov::Tensor` is create as "dynamic", `set_shape` will allow change to any shape. This is already used in [softsign evaluate](https://github.com/openvinotoolkit/openvino/blob/27ea9eab329cd7db547f5aeb24bfca7e549ef489/src/core/src/op/softsign.cpp#L95), and there is a case when this evaluate is called with "dynamic" output tensor and reshaped.

Maybe `set_shape` WA should be improved?

Comment (maxnick) [reply src/core/src/tensor_conversion_util.cpp]: It looks like I mixed up `ov::Tensor` and `runtime::Tensor`, since the letter being initialized from "dynamic" `ov::Shape` can not perform `set_shape` call, because an exception is thrown. But, `ov::Tensor` performs `set_shape` properly, even though being instantiated from the result of the `ov::util::make_dynamic_shape()` call.  So I adopted the `ov::util::make_dynamic_shape()` utility on the plugin level.

File: src/core/dev_api/shape_util.hpp [hunk 6]
```diff
@@ -25,6 +25,6 @@ Shape make_dynamic_shape();
  * @return True if shape is dynamic otherwise false.
  */
 OPENVINO_DEPRECATED("This function is deprecated and will be removed soon.")
-bool is_dynamic_shape(const Shape& s);
+OPENVINO_API bool is_dynamic_shape(const Shape& s);
```
Comment (praasz) [inline src/core/dev_api/shape_util.hpp:28]: can you check if export of this required. I think is used in core only. 
If not required maybe is better not export as also marked as deprecated.

**Review by praasz — APPROVED** [review]

File: src/core/dev_api/shape_util.hpp (hunk 6 above)
Comment (maxnick) [reply src/core/dev_api/shape_util.hpp:28]: It is used by the CPU plugin as well, so this is the reason why they are exported. I also though about making these utils header only, but the usage of a static variable changed my mind.

**Comment by maxnick** [comment]
@ilyachur , do you agree with moving the shape utils to the dev api?

**Comment by maxnick** [comment]
@dmitry-gorokhov , could you please review?

**Review by ilyachur — APPROVED** [review]
Core part LGTM

**Review by dmitry-gorokhov — APPROVED** [review]
Would be nice to provide a little bit broader explanation for "skip" status and use-cases there it is expected to be applied.

**Comment by maxnick** [comment]
> Would be nice to provide a little bit broader explanation for "skip" status and use-cases there it is expected to be applied.

@dmitry-gorokhov , the code comments have been extended.
